File: books/cs/general-purpose-algorithms/introduction-to-algorithms2/chapter8/PHP/counting_sort.php
<?php

define("DATASET_SIZE", 100);

$lower = 0;
$upper = 100;
$k = 0;
$aaa;
$bbb;
$ccc;
$idx;
$total;

$random_number = mt_rand($lower,$upper);
for( $idx=0; $idx<DATASET_SIZE; $idx++){
	$aaa[$idx] = mt_rand($lower,$upper);
	$bbb[$idx] = 0;
}
print("the start of array aaa.\n");
var_dump( $aaa );
print("the end of array aaa.\n\n\n");

$k = max($aaa);
for( $idx=0; $idx<=$k; $idx++ ){
	$ccc[$idx] = 0;
}
for( $idx=0; $idx<DATASET_SIZE; $idx++){
	$ccc[$aaa[$idx]] += 1;
}
$total=0;
for( $idx=0; $idx<=$k; $idx++ ){
	$total += $ccc[$idx];
	$ccc[$idx] = $total;
}

print("the start of array ccc.\n");
var_dump( $ccc );
print("the end of array ccc.\n\n\n");

for( $idx=DATASET_SIZE-1; $idx>=0; $idx--){
	$bbb[$ccc[$aaa[$idx]]] = $aaa[$idx];
	$ccc[$aaa[$idx]] -= 1;
}
print("the start of array bbb.\n");
var_dump( $bbb );
print("the end of array bbb.\n\n\n");
?>

File: books/cs/general-purpose-algorithms/introduction-to-algorithms2/chapter8/Swift/counting_sort.swift
/**
* counting sort algorithm
* introduction to algorithm, 3rd edition, chapter 3
*
*/

import Darwin
import Cocoa

let dataset_size = 10
let max: UInt32 = 10
var aaa:[Int] = [Int]()
var bbb:[Int] = [Int]()
var ccc:[Int] = [Int]()
var total:Int = 0

for index in 0 ... (dataset_size-1) {
	aaa.append( Int(arc4random_uniform(max)) )
	print("index = \(index), value = \(aaa[index])\n")
}

let k = aaa.maxElement()
print(k)
ccc = Array(count:(k!+1), repeatedValue:0)
bbb = Array(count:(dataset_size+1),repeatedValue:0)

for index in aaa {
	ccc[index] += 1
	print("value = \(index)\n")
}
for index in 0...k! {
	total += ccc[index]
	print("index = \(index), value = \(ccc[index]), total = \(total)\n")
	ccc[index] = total
}

for index in aaa.reverse() {
	bbb[ccc[index]] = index
	ccc[index] -= 1
}
for index in 0 ... dataset_size {
	print("index = \(index), value = \(bbb[index])\n")
}
